common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -339,10 +339,11 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
    * @return the uppercase string
    */
   public static UTF8String toUpperCase(final UTF8String target) {
-    return UTF8String.fromString(toUpperCase(target.toString()));
+    if (target.isFullAscii()) return target.toUpperCaseAscii();
+    return UTF8String.fromString(toUpperCaseSlow(target.toString()));
```

Comment:
```suggestion
    return UTF8String.fromString(toUpperCaseSlow(target.makeValid().toString()));
```
I think we need to do the illegal UTF8 sequence replacement with `makeValid` as opposed to rely on Java.

Reply:
I think there's plenty of places where we need to do `makeValid`, so I suggest the following: let's first go with these changes, and leave the "replacement injection" for a separate PR - in that PR, we should carefully use makeValid `wherever` needed and do testing with special cases to verify the behaviour

how does this sound?

Reply:
Agreed. But let's not postpone this for too longer please.